packaging/os/xbps.py

```diff
@@ -0,0 +1,301 @@
+                (state == "latest" and not updated)):
+            would_be_changed.append(package)
+    if would_be_changed:
+        if state == "absent":
+            state = "removed"
+        module.exit_json(changed=True, msg="%s package(s) would be %s" % (
+            len(would_be_changed), state))
+    else:
+        module.exit_json(changed=False, msg="package(s) already %s" % state)
+
+
+def main():
+    """Returns, calling appropriate command"""
+
+    mask = os.umask(0)  # fix the umask
+
+    module = AnsibleModule(
+        argument_spec=dict(
+            name=dict(default=None, aliases=['pkg', 'package'], type='list'),
+            state=dict(default='present', choices=['present', 'installed',
+                                                   "latest", 'absent',
+                                                   'removed']),
+            recurse=dict(default=False, type='bool'),
+            force=dict(default=False, type='bool'),
+            upgrade=dict(default=False, type='bool'),
+            update_cache=dict(default=True, aliases=['update-cache'],
+                              type='bool')
+        ),
+        required_one_of=[['name', 'update_cache', 'upgrade']],
+        supports_check_mode=True)
+
+    xbps_path = dict()
+    xbps_path['install'] = module.get_bin_path('xbps-install', True)
+    xbps_path['query'] = module.get_bin_path('xbps-query', True)
+    xbps_path['remove'] = module.get_bin_path('xbps-remove', True)
+
+    if not os.path.exists(xbps_path['install']):
+        module.fail_json(msg="cannot find xbps, in path %s"
+                         % (xbps_path['install']))
+
+    p = module.params
+
+    # normalize the state parameter
+    if p['state'] in ['present', 'installed']:
+        p['state'] = 'present'
+    elif p['state'] in ['absent', 'removed']:
+        p['state'] = 'absent'
+
+    if p["update_cache"] and not module.check_mode:
+        changed = update_package_db(module, xbps_path)
+        if p['name'] is None and not p['upgrade']:
+            if changed:
+                module.exit_json(changed=True,
+                                 msg='Updated the package master lists')
+            else:
+                module.exit_json(changed=False,
+                                 msg='Package list already up to date')
+
+    if p['update_cache'] and module.check_mode\
+       and not (p['name'] or p['upgrade']):
+        module.exit_json(changed=True,
+                         msg='Would have updated the package cache')
+
+    if p['upgrade']:
+        upgrade(module, xbps_path)
+
+    if p['name']:
+        pkgs = p['name']
+
+        if module.check_mode:
+            check_packages(module, xbps_path, pkgs, p['state'])
+
+        if p['state'] in ['present', 'latest']:
+            install_packages(module, xbps_path, p['state'], pkgs)
+        elif p['state'] == 'absent':
+            remove_packages(module, xbps_path, pkgs)
+
+    os.umask(mask)  # Reset the umask to original value
```

Comment:
Also, this has to be a context manager, since `module.exit_json` and `.fail_json` both end the execution of the module, so the correct umask will never get restored.